lib/components/user/notification-prefs-pane.js

```diff
@@ -85,28 +146,236 @@ const NotificationPrefsPane = ({
-          </FormGroup>
+          <Formik
+            validateOnChange
+            validationSchema={codeValidationSchema}
+            initialValues={initialFormikValues}
+          >
+            {
+              // Pass Formik props to the component rendered so Formik can manage its validation.
+              // (The validation for this component is independent of the validation set in UserAccountScreen.)
+              innerProps => {
+                return (
+                  <PhoneNumberEditor
+                    {...innerProps}
+                    initialPhoneNumber={phoneNumber}
+                    initialPhoneNumberVerified={isPhoneNumberVerified}
+                    onRequestCode={onRequestPhoneVerificationCode}
+                    onSubmitCode={onSendPhoneVerificationCode}
+                    phoneFormatOptions={phoneFormatOptions}
+                  />
+                )
+              }
+            }
+          </Formik>
         )}
       </Details>
     </div>
   )
 }
 
 export default NotificationPrefsPane
+
+/**
+ * Sub-component that handles phone number and validation code editing and validation intricacies.
+ */
+class PhoneNumberEditor extends Component {
+  constructor (props) {
+    super(props)
+
+    const { initialPhoneNumber } = props
+    this.state = {
+      // If true, phone number is being edited.
+      // For new users, render component in editing state.
+      isEditing: isBlank(initialPhoneNumber),
+
+      // Holds the new phone number (+15555550123 format) entered by the user
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
+  }
+
+  _handleEditNumber = () => {
+    this.setState({
+      isEditing: true,
+      newPhoneNumber: ''
+    })
+  }
+
+  _handleNewPhoneNumberChange = newPhoneNumber => {
+    this.setState({
+      newPhoneNumber
+    })
+  }
+
+  _handleCancelEditNumber = () => {
+    this.setState({
+      isEditing: false
+    })
+  }
+
+  _handlePhoneNumberKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the phone number field,
+      // prevent form submission and request the code.
+      e.preventDefault()
+      this._handleRequestCode()
+    }
+  }
+
+  _handleValidationCodeKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the validation code field,
+      // prevent form submission and send the validation code.
+      e.preventDefault()
+      this._handleSubmitCode()
+    }
+  }
+
+  /**
+   * Send phone verification request with the entered values.
+   */
+  _handleRequestCode = () => {
+    const { initialPhoneNumber, initialPhoneNumberVerified, onRequestCode } = this.props
+    const { newPhoneNumber } = this.state
+
+    this._handleCancelEditNumber()
+
+    // Send the sms request, unless the user
+    // entered the same phone number as their current verified number.
```

Comment:
These comments could have better line breaks.

Reply:
Addressed in https://github.com/opentripplanner/otp-react-redux/pull/224/commits/cbf2d86940c5b4c51d9eb03969c22926c1a2ba7b.